rfd/0169-auto-updates-linux-agents.md

```diff
@@ -0,0 +1,364 @@
+---
+authors: Stephen Levine (stephen.levine@goteleport.com)
+state: draft
+---
+
+# RFD 0169 - Automatic Updates for Linux Agents
+
+## Required Approvers
+
+* Engineering: @russjones && @bernardjkim
+* Product: @klizhentas || @xinding33 
+* Security: @reedloden
+
+## What
+
+This RFD proposes a new mechanism for Teleport agents installed on Linux servers to automatically update to a version set by an operator via tctl.
+
+The following anti-goals are out-of-scope for this proposal, but will be addressed in future RFDs:
+- Analogous adjustments for Teleport agents installed on Kubernetes
+- Phased rollouts of new agent versions for agents connected to an existing cluster
+- Signing of agent artifacts via TUF
+- Teleport Cloud APIs for updating agents
+
+This RFD proposes a specific implementation of several sections in https://github.com/gravitational/teleport/pull/39217.
+
+Additionally, this RFD parallels the auto-update functionality for client tools proposed in https://github.com/gravitational/teleport/pull/39805.
+
+## Why
+
+The existing mechanism for automatic agent updates does not provide a hands-off experience for all Teleport users.
+
+1. The use of system package management leads to interactions with `apt upgrade`, `yum upgrade`, etc. that can result in unintentional upgrades or confusing command output.
+2. The use of system package management requires complex logic for each target distribution.
+3. The installation mechanism requires 4-5 commands, includes manually installing multiple packages, and varies depending on your version and edition of Teleport.
+4. The use of bash to implement the updater makes changes difficult and prone to error.
+5. The existing auto-updater has limited automated testing.
+6. The use of GPG keys in system package managers has key management implications that we would prefer to solve with TUF in the future.
+7. The desired agent version cannot be set via Teleport's operator-targeted CLI (tctl).
+8. The rollout plan for the new agent version is not fully-configurable using tctl.
+9. Agent installation logic is spread between the auto-updater script, install script, auto-discovery script, and documentation.
+10. Teleport contains logic that is specific to Teleport Cloud upgrade workflows.
+11. The existing auto-updater is not self-updating.
+12. It is difficult and undocumented to automate agent upgrades with custom automation (e.g., with JamF). 
+
+We must provide a seamless, hands-off experience for auto-updates that is easy to maintain.
+
+## Details
+
+We will ship a new auto-updater package written in Go that does not interface with the system package manager.
+It will be versioned separately from Teleport, and manage the installation of the correct Teleport agent version manually. 
```

Comment:
Where will the code for tool live? I'm strongly of the opinion that if it's not versioned with Teleport then it shouldn't be in this repo or teleport.e, rather, it should be in it's own repo.

Reply:
The `teleport-updater` binary would also ship in the Teleport tarball, so it could be versioned with Teleport.

That said, it might nice if the `teleport-upgrader` package was versioned separately from Teleport, since the package version won't match the version of Teleport installed (or the version of the updater that is updated by the updater, after the initial Teleport update).

I could see this working either way. No strong opinion.

Reply:
I'm confused here - the RFD says:
> It will be versioned separately from Teleport, and manage the installation of the correct Teleport agent version manually.

But here you've said:
> The teleport-updater binary would also ship in the Teleport tarball

and

> it **might** nice if the teleport-upgrader package was versioned separately from Teleport

Is the tool going to be versioned with Teleport or not? If not, and if it'll be used to install Teleport, then why would it ship inside a Teleport tarball?

Reply:
@russjones is there a benefit to keeping this versioned with Teleport? This has caused problems in the past (in fact I spent hours today working on a problem caused by this), and there is precedence for not versioning customer-facing tools/products with Teleport (see shared-workflows tools, the projects we've forked, and TAG).

Here are the benefits I see of pulling this code out of teleport/teleport.e:
* It cuts down on release time. If we pulled this out today then every Teleport releases would take 15-30m less.
* It keeps the code from being accidentally coupled with Teleport code. This has bitten me several times in the past two weeks alone.
* It allows us to release updater changes separately, and makes it more clear to customers when something changes. Updater code has historically changed very infrequently, however customers think that it's changed on every release because we release it with Teleport.
* Changes to the updater will not need to be backported, unless they are specifically needed for an older version of the updater. The subtle but important difference is that because the updater would likely have few major releases, this would not need to be done for every change.
* It integrates well with the security model that our tools (i.e. GitHub) uses. A compromise to this code in a separate repo has a much smaller impact to Gravitational than if there was a compromise in teleport/teleport.e, as the GH Environments in these repos have access to do significantly more damage.
* It would allow for faster iteration for both the tool itself, and Teleport. More code in the teleport/teleport.e code bases results in more merge contention (which also indirectly results in increased CI costs).

Reply:
Hey folks this discussion somehow disappeared, so I'm commenting here in hopes that it shows back up.